server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,703 @@
+ * compatible open source license.
+ */
+
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.gateway;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.action.support.nodes.BaseNodesResponse;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.cluster.routing.RerouteService;
+import org.opensearch.cluster.routing.RoutingNodes;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.AllocateUnassignedDecision;
+import org.opensearch.cluster.routing.allocation.FailedShard;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.Priority;
+import org.opensearch.common.UUIDs;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lease.Releasables;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.common.util.set.Sets;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.indices.store.ShardAttributes;
+import org.opensearch.indices.store.TransportNodesListShardStoreMetadata;
+import org.opensearch.indices.store.TransportNodesListShardStoreMetadataBatch;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Spliterators;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Allocator for the Shards batch gateway
+ *
+ * @opensearch.internal
+ */
+public class ShardsBatchGatewayAllocator extends GatewayAllocator {
```

Comment:
Can you explain your reasoning for  preferring `ShardsBatchGatewayAllocator extends GatewayAllocator` over `ShardsBatchGatewayAllocator implements ExistingShardsAllocator ` ?

As per my understanding, methods in `GatewayAllocator` are incompatible with batch mode and can/will fail if invoked by batch allocation object. In such a scenario, having the `ShardsBatchGatewayAllocator` implement the `ExistingShardsAllocator` interface guarantees that all methods are context-aware of batch operation.

Reply:
For two reasons-
1. ShardBatchGatewayAllocator is of type GatewayAllocator only. It does the same thing what GatewayAllocator does but in batch mode, so in object oriented sense it makes sense to extend it
2. ExistingShardsAllocator have `allocateUnassigned` non default method, which needs to be implemented by this batched version of GA(ShardBatchGatewayAllocator). So in that case we need to throw IllegalArgumentException. Now think of case of many indices in the cluster which have different Allocators set, then in that case we dont want batched version of GA(ShardBatchGatewayAllocator) to take responsibility rather non-bacthed version of GA to take responbility, then the extension of classes will help to fallback to non-batched version of GA(that is GatewayAllocator itself)

Reply:
1. Interfaces in Java is a tool to enforce a contract, while Inheritance is to re-use the methods and fields of Parent Class. Is `ShardBatchGatewayAllocator` reusing any fields or methods from `GatewayAllocator` ?
2. For handling Indices with Different Allocator types, We are already instantiating and [passing the `GatewayAllocator` and `ShardBatchGatewayAllocator` object](https://github.com/opensearch-project/OpenSearch/pull/8746/files#diff-f839350640f8bd27f602b3249215252bf07afef5d7e480b23271788af6cda331R1194-R1200), the decision-making code should invoke the appropriate method for the allocator object:

```
ExistingShardsAllocator allocator = getAndVerifySameAllocatorForAllUnassignedShards(allocation);
            if (allocator != null) {
                // use batch mode implementation of GatewayAllocator
                ...
                shardsBatchGatewayAllocator.allocateUnassignedBatch(allocation, true);
                ...
            } else {
                // it means though batch mode is enabled but some indices have custom allocator set, execute per-shard assignment
                 ...
                gatewayAllocator.allocateUnassigned(shardRouting, allocation, iterator);
                 ...
            }

````